### PR TITLE
feat(android): Feature/keyboard download manager

### LINKED ISSUE #1316: [Android] Move keyboard updates into a background async process
From PR #1301 @mcdurdin writes:
> Long term we need to move updates into a background async process that uses
system notifications or similar to let user accept them with minimum of
interference.

--------
Some possibilities with the Android app

* Have the app icon show a notification dot when a keyboard update is available
* In the keyboards list, have some way to notify which keyboard update is available

### PR DESCRIPTION
**Pull request by safolit**
Fixes #1316

## Previous Behavior ##
Keyboard download in keyman android could take lot of time. Currently the user has to wait until download is finished because there is a modal dialog showing a infinite spinner.

## New Behavior ##
The Keyman keyboard is downloaded using the background DownloadManager using the same approach as the catalog download.
The Download happens in the background and the user can continue to use the application during the download. The user gets notified via a Android toast when the download is running and when it is finished.
The new keyboard is not automatically selected. The new Keyboard is marked with "[new]" in keyboard selection dialog.

Start Downloading new Keyboard from catalog
Screenshot_1572505655

Confirm download
Screenshot_1572505660

Download is started in background notification
Screenshot_1572505663

Try to install the same keyboard again:
Screenshot_1572505675

Notification that download is finished
Screenshot_1572505702

Additional feature: Mark don't switch keyboard when download is completed. Add new keyboard marker
Screenshot_1572862339

After selecting the keyboard the first time, new flag is removed.
Screenshot_1572862354

## Work Done ##
- Start cloud download from KeyboardDownloadActivity
-  Add Cloud download Callbacks for Keyboardmetadata, keyboard package, fonts and dictionary
-  Flow chart for processes: This pull request contains processes concerning keyboard and dictionary/ lexical model update
[keyboard_download_diagram.pdf](https://github.com/keymanapp/keyman/files/3792080/keyboard_download_diagram.pdf)

## Next Steps ##
- Get meta data from keyboard cache
- Review notifications
- Remove old Downloadtask

## Testing 
- Goto Settings/ Installed languages
- Add a new language
- Select a keyboard from the list
- Download will start in background and send a Notification
- When download is completed, a Notification comes up, telling the resources are up to date.
- List of installed languages should show the new language
- Go back to the text view -- New Keyboard is not selected automatically
- Goto keyboard selection dialog
- the new keyboard is marked with "[new]"
- select the new keyboard
- reopen the keyboard selection dialog --> Keyboard is not marked as new

### CODE REVIEW

**Comment by darcywong00** [comment]
This branch looks like it's based off of #2247, so reviewers should do that one first.

**Comment by mcdurdin** [comment]
One issue with this model is that the keyboard download normally shows a welcome screen after installation to help the user get started. We need to find a way to make this work smoothly in this new UX paradigm.

**Comment by darcywong00** [comment]
> One issue with this model is that the keyboard download normally shows a welcome screen after installation to help the user get started. We need to find a way to make this work smoothly in this new UX paradigm.

Yes, let's discuss the workflow at planning. Currently, we're not pulling welcome.htm files from the legacy `https://api.keyman.com/cloud/4.0/languages` endpoint. *But* when we eventually change keyboard downloads to pull down the .kmp files via Download Manager, we'll need a design for displaying the readme.htm and welcome.htm files.

**Comment by safolit** [comment]
I think this welcome screen is only shown for custom keyboards. Isn't it?

**Comment by darcywong00** [comment]
Since #2247 has merged to master, can you merge master into this branch and push your branch?

**Comment by darcywong00** [comment]
> I am concerned that we have a decrease in usability due to a slightly less obvious process for users, as background install may make them think the keyboard install isn't working. 

Recap of team discussion.
1. Since currently the `welcome.htm` isn't pulled down on keyboard downloads, we won't change that behavior.
2. Do not change to new keyboard when it finished downloading. Just keep the current keyboard until user selects the new one.
3. For clarity, we'll just note the new keyboard(s) in the keyboards list by appending a string like ` (new)`. After they've selected a new keyboard, we'll remove the addendum.

**Comment by safolit** [comment]
@darcywong00 Merge master to keyboard-download-manager branch is done

File: android/KMEA/app/src/main/java/com/tavultesoft/kmea/data/CloudKeyboardMetaDataDownloadCallback.java [hunk 1]
```diff
@@ -0,0 +1,349 @@
...
+
+          CloudDownloadMgr.getInstance().executeAsDownload(aContext, _r.additionalDownloadid, null, _callback,
+            _r.additionalDownloads.toArray(new CloudApiTypes.CloudApiParam[0]));
+        }
+        else if(_r.returnjson.target== CloudApiTypes.ApiTarget.KeyBoardLexicalModels) {
+
+          if(  CloudDownloadMgr.getInstance().alreadyDownloadingData(_r.additionalDownloadid))
+          {
+            Toast.makeText(aContext,
+              aContext.getString(R.string.dictionary_download_is_running_in_background),
```
Comment (darcywong00) [inline android/KMEA/app/src/main/java/com/tavultesoft/kmea/data/CloudKeyboardMetaDataDownloadCallback.java]: When I install the **Salish, Straights --> SENCOTEN** keyboard I see this toast notification.

But I'm not seeing nrc.str.sencoten.model in the file system (appdata/models).
I would have expected the dictionary to also download and be enabled when I switch to the Sencoten keyboard.

Comment (darcywong00) [reply android/KMEA/app/src/main/java/com/tavultesoft/kmea/data/CloudKeyboardMetaDataDownloadCallback.java]: Fixed in https://github.com/keymanapp/keyman/pull/2275/commits/b9765970750fbe8461adcd7279298dfca01257a2

File: android/KMEA/app/src/main/java/com/tavultesoft/kmea/data/CloudApiTypes.java [hunk 2]
```diff
@@ -33,25 +35,44 @@ public CloudApiReturns(ApiTarget target, JSONObject jsonObject) {
...
 
-  protected enum ApiTarget {
+  public enum ApiTarget {
     Keyboards,
-    LexicalModels
+    LexicalModels,
+    Keyboard,
+    KeyBoardLexicalModels,
+    KeyboardData,
+    LexicalModelPackage,
```
Comment (darcywong00) [inline android/KMEA/app/src/main/java/com/tavultesoft/kmea/data/CloudApiTypes.java]: 1. Can you lower-case the 'b' in `KeyBoardLexicalModels`?

2. Please add some notes how these ApiTargets get used: 
(for instance)
* Keyboards vs Keyboard
* LexicalModels vs KeyBoardLexicalModels vs LexicalModelPackage

File: android/KMEA/app/src/main/java/com/tavultesoft/kmea/data/CloudDataJsonUtil.java [hunk 3]
```diff
@@ -29,6 +31,27 @@ private CloudDataJsonUtil()
   {
     //no instances
   }
+
+  public static HashMap<String,String> createKeyBoardInfoMap(String aPackageId,String aLanguageId, String aLanguageName, String aKeyboardId,
```
Comment (darcywong00) [inline android/KMEA/app/src/main/java/com/tavultesoft/kmea/data/CloudDataJsonUtil.java]: Let's rename `createKeyBoardInfoMap` to `createKeyboardInfoMap` too.

File: android/KMEA/app/src/main/java/com/tavultesoft/kmea/data/CloudDataJsonUtil.java [hunk 4]
```diff
@@ -220,4 +237,187 @@ public static File getLexicalModelCacheFile(Context context) {
...
+        }
+      } catch (JSONException e) {
+        return null;
+      }
+    }
+
+    return urls;
+  }
+
+  public static JSONObject findMatchingKeyBoardByID(JSONArray aKeyboards, String aKbId)
```
Comment (darcywong00) [inline android/KMEA/app/src/main/java/com/tavultesoft/kmea/data/CloudDataJsonUtil.java]: Another one to rename: `findMatchingKeyBoardByID`. Might as well replace all instances of `KeyBoard` with `Keyboard`.

**Pull request merged**